### PR TITLE
utils: use enumerations to locate build locations

### PR DESCRIPTION
**Pull request by compnerd**
Rather than hardcode the build numbers, use named parameters to locate them during the build process.

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @hyp @hjyamauchi @tristanlabelle

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Review by hjyamauchi — APPROVED** [review]

File: utils/build.ps1 [hunk 1]
```diff
@@ -1474,25 +1507,25 @@ function Build-LLBuild($Arch, [switch]$Test = $false) {
...
       Isolate-EnvVars {
         Invoke-VsDevShell $HostArch
-        Invoke-Program ninja.exe -C "$BinaryCache\0" FileCheck not
+        Invoke-Program ninja.exe -C (Get-HostProjectBinaryCache BuildTools) FileCheck not
       }
 
       $Targets = @("default", "test-llbuild")
       $TestingDefines = @{
-        FILECHECK_EXECUTABLE = "$BinaryCache\0\bin\FileCheck.exe";
+        FILECHECK_EXECUTABLE = (Join-Path -Path (Join-Path -Path (Get-HostProjectBinaryCache BuildTools) -ChildPath "bin") -ChildPath "FileCheck.exe");
```
Comment (tristanlabelle) [inline utils/build.ps1:1515]: The `Join-Path` seem a little overkill throughout?
```suggestion
        FILECHECK_EXECUTABLE = "$(Get-HostProjectBinaryCache BuildTools)\bin\FileCheck.exe";
```

File: utils/build.ps1 [hunk 2]
```diff
@@ -254,16 +254,47 @@ function Invoke-BuildStep([string]$Name) {
...
   XCTest
 }
 
-function Get-ProjectBinaryCache($Arch, [Target]$Target) {
-  return "$BinaryCache\" + ($Arch.BuildID + $Target.value__)
+function Get-TargetProjectBinaryCache($Arch, [TargetComponent]$Project) {
+  return "$BinaryCache\" + ($Arch.BuildID + $Project.value__)
+}
+
+enum HostComponent {
```
[the remark this answers is not in the document]
Comment (compnerd) [reply utils/build.ps1:269]: No, the numbers should be left in tact (but, the beauty is that we don't need to synchronize those after this - everything should be left in a consistent state.

File: utils/build.ps1 (hunk 1 above)
Comment (compnerd) [reply utils/build.ps1:1515]: Would be nice to be able to support other platforms some day (e.g. Linux)

File: utils/build.ps1 [hunk 3]
```diff
@@ -254,16 +254,47 @@ function Invoke-BuildStep([string]$Name) {
...
+  SymbolKit
+  DocC
+}
+
+function Get-HostProjectBinaryCache([HostComponent]$Project) {
+  return "$BinaryCache\$($Project.value__)"
+}
+
+function Get-HostProjectCMakeModules([HostComponent]$Project) {
+  return "$BinaryCache\$($Project.value__)\cmake\modules"
```
Comment (tristanlabelle) [inline utils/build.ps1:297]: All remaining instances of hard-coded values in this file are to `$BinaryCache\1` or `$BinaryCache\7`, so how about defining two globals after this function definition:

```ps1
$CompilersBinaryCache = Get-HostProjectBinaryCache Compilers
$DriverBinaryCache = Get-HostProjectBinaryCache Driver

File: utils/build.ps1 [hunk 4]
```diff
@@ -1601,15 +1634,15 @@ function Build-ASN1($Arch) {
...
+    -Bin (Get-HostProjectBinaryCache Certificates) `
     -Arch $Arch `
     -UseBuiltCompilers Swift `
     -SwiftSDK $SDKInstallRoot `
     -BuildTargets default `
     -Defines @{
       BUILD_SHARED_LIBS = "NO";
-      SwiftASN1_DIR = "$BinaryCache\10\cmake\modules";
-      SwiftCrypto_DIR = "$BinaryCache\8\cmake\modules";
+      SwiftCrypto_DIR = (Get-HostProjectCMakeModules Crypto);
```
Comment (tristanlabelle) [inline utils/build.ps1:1644]: Parens are probably not necessary on the RHS of an assignment.

File: utils/build.ps1 [hunk 5]
```diff
@@ -254,16 +254,47 @@ function Invoke-BuildStep([string]$Name) {
...
   Runtime
   Dispatch
   Foundation
   XCTest
 }
 
-function Get-ProjectBinaryCache($Arch, [Target]$Target) {
-  return "$BinaryCache\" + ($Arch.BuildID + $Target.value__)
+function Get-TargetProjectBinaryCache($Arch, [TargetComponent]$Project) {
+  return "$BinaryCache\" + ($Arch.BuildID + $Project.value__)
```
Comment (tristanlabelle) [inline utils/build.ps1:266]: Oh wow, `value__` is the name of the hidden field that .NET generates in structs derived from `System.Enum` per standard. Cool trick!

**Review by tristanlabelle — APPROVED** [review]
Love it!

File: utils/build.ps1 (hunk 4 above)
Comment (compnerd) [reply utils/build.ps1:1644]: Oh, it shouldn't confuse things to have `Get-HostProjectCMakeModules Crypto;`? I definitely would prefer to remove the parenthesis then.

File: utils/build.ps1 (hunk 3 above)
Comment (tristanlabelle) [reply utils/build.ps1:297]: I'm not sure if the usages are in the same function, I was suggesting creating global variables for them. (but even better if they can be locals)

File: utils/build.ps1 (hunk 1 above)
Comment (tristanlabelle) [reply utils/build.ps1:1515]: We could create a Join-Path variant that takes an arbitrary number of components and doesn't require labels so it's tidier.

Comment (tristanlabelle) [reply utils/build.ps1:1515]: or use `[IO.Path]::Combine($foo, $bar, $baz)`

Comment (compnerd) [reply utils/build.ps1:1515]: Let me try switching to `[IO.Path]::Combine` ... if we can use builtins, I'd prefer that.

File: utils/build.ps1 (hunk 4 above)
Comment (compnerd) [reply utils/build.ps1:1644]: Seems that they are needed :( - maybe you can try to clean them up subsequently? I'd love to get that noise removed.

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Pull request merged**